src/dbnode/storage/series/buffer.go

```diff
@@ -268,65 +385,106 @@ func bucketTick(now time.Time, b *dbBuffer, idx int, start time.Time) int {
 
 func (b *dbBuffer) DrainAndReset() drainAndResetResult {
 	// Avoid capturing any variables with callback
-	mergedOutOfOrder := b.computedForEachBucketAsc(computeAndResetBucketIdx, bucketDrainAndReset)
+	mergedOutOfOrder := b.computedForEachBucketAsc(computeAndResetBucketIdx, bucketTypeRealTime, bucketDrainAndReset)
 	return drainAndResetResult{
 		mergedOutOfOrderBlocks: mergedOutOfOrder,
 	}
 }
 
-func bucketDrainAndReset(now time.Time, b *dbBuffer, idx int, start time.Time) int {
+func bucketDrainAndReset(now time.Time, b *dbBuffer, id bucketID, start time.Time) int {
 	mergedOutOfOrderBlocks := 0
+	bucket := b.bucketAtIdx(id)
 
-	if b.buckets[idx].needsDrain(now, start) {
-		// Rotate the buffer to a block, merging if required
-		result, err := b.buckets[idx].discardMerged()
-		if err != nil {
+	if bucket.needsDrain(now, start) {
+		mergedOutOfOrderBlocks += b.bucketDrain(now, id, start)
+	}
+
+	if bucket.needsReset(now, start) {
+		bucket.resetTo(start, true)
+	}
+
+	if bucket.isStale(now) {
+		b.removeBucket(id.key)
+	}
+
+	return mergedOutOfOrderBlocks
+}
+
+func (b *dbBuffer) bucketDrain(now time.Time, id bucketID, start time.Time) int {
+	mergedOutOfOrderBlocks := 0
+
+	bucket := b.bucketAtIdx(id)
+	// Rotate the buffer to a block, merging if required
+	result, err := bucket.discardMerged()
+	if err != nil {
+		log := b.opts.InstrumentOptions().Logger()
+		log.Errorf("buffer merge encode error: %v", err)
+	} else {
+		if result.merges > 0 {
+			mergedOutOfOrderBlocks++
+		}
+		if !(result.block.Len() > 0) {
 			log := b.opts.InstrumentOptions().Logger()
-			log.Errorf("buffer merge encode error: %v", err)
+			log.Errorf("buffer drain tried to drain empty stream for bucket: %v",
+				start.String())
 		} else {
-			if result.merges > 0 {
-				mergedOutOfOrderBlocks++
+			// If this block was read mark it as such
+			if lastRead := bucket.lastRead(); !lastRead.IsZero() {
+				result.block.SetLastReadTime(lastRead)
 			}
-			if !(result.block.Len() > 0) {
-				log := b.opts.InstrumentOptions().Logger()
-				log.Errorf("buffer drain tried to drain empty stream for bucket: %v",
-					start.String())
-			} else {
-				// If this block was read mark it as such
-				if lastRead := b.buckets[idx].lastRead(); !lastRead.IsZero() {
-					result.block.SetLastReadTime(lastRead)
-				}
+			if b.drainFn != nil {
 				b.drainFn(result.block)
 			}
 		}
-
-		b.buckets[idx].drained = true
 	}
 
-	if b.buckets[idx].needsReset(start) {
-		// Reset bucket
-		b.buckets[idx].resetTo(start)
+	bucket.drained = true
+	bucket.setLastWrite(now)
+	bucket.resetNumWrites()
+	if bucket.isStale(now) {
```

Comment:
confused about the stale chck and removal again. Once its drained, why not just remove it? Also since you just called `setLastWrite(now)` wont it always be false?

Reply:
Agreed - this logic was a little convoluted. Cleaned this up to decouple this and only have bucketTick deal with stale bucket removal.